src/html-imports.js

```diff
@@ -9,718 +9,766 @@
+        waitForStyles(this._doc)
+      ]).then(() => fireEvents(this._doc));
+    }
+
+    _flatten(element) {
+      const n$ = element.querySelectorAll(IMPORT_SELECTOR);
+      for (let i = 0, l = n$.length, n; i < l && (n = n$[i]); i++) {
+        n.import = this.documents[n.href];
+        if (n.import && !n.import.__firstImport) {
+          n.import.__firstImport = n;
+          this._flatten(n.import);
+          // If in the main document, observe for any imports added later.
+          if (element === document) {
+            // In IE/Edge, when imports have link stylesheets/styles, the cascading order
+            // isn't respected https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/10472273/
+            if (isIE || isEdge) {
+              cloneAndMoveStyles(n);
+            }
+            this._observe(n.import);
           }
+          n.appendChild(n.import);
         }
       }
     }
-  },
-
-// TODO(valdrin): if possible, fire in same order as native imports which is
-// probably bottom up.
-  _fireEvents: function(element) {
-    var n$ = element.querySelectorAll(IMPORT_SELECTOR);
-    for (var i=0; i < n$.length; i++) {
-      var n = n$[i];
-      if (!n.__loaded) {
-        flags.log && console.warn('fire', n.import ? 'load' : 'error', n.href);
-        n.__loaded = true;
-        n.dispatchEvent(new CustomEvent(n.import ? 'load' : 'error'));
+
+    _observe(element) {
+      if (element.__importObserver) {
+        return;
       }
+      element.__importObserver = new MutationObserver(this._onMutation.bind(this));
+      element.__importObserver.observe(element, {
+        childList: true,
+        subtree: true
+      });
     }
-  },
-
-  observe: function(element) {
-    if (!element.__importObserver) {
-      element.__importObserver = new MutationObserver(function (mxns) {
-        mxns.forEach(function(m) {
-          if (m.addedNodes) {
-            for (var i=0; i < m.addedNodes.length; i++) {
-              var p = m.addedNodes[i];
-              // TODO(sorvell): x-platform matches
-              if (p.nodeType === Node.ELEMENT_NODE &&
-                  p.matches(IMPORT_SELECTOR)) {
-                importer.loadNode(p);
-              }
+
+    /**
+     * @param {Array<MutationRecord>} mutations
+     */
+    _onMutation(mutations) {
+      const promises = [];
+      for (let j = 0, m; j < mutations.length && (m = mutations[j]); j++) {
+        for (let i = 0, l = m.addedNodes ? m.addedNodes.length : 0; i < l; i++) {
+          const n = /** @type {Element} */ (m.addedNodes[i]);
+          if (n && isImportLink(n)) {
+            promises.push(whenElementLoaded(n));
+            if (!useNative) {
+              this._loader.addNode(n);
             }
           }
-        });
-      });
-      element.__importObserver.observe(element, {childList: true, subtree: true});
+        }
+      }
+      if (promises.length) {
+        // Ensure we update isImporting.
+        isImporting = true;
```

Comment:
This technique will not work when `whenReady` is used before the mutation observer notices the import. For example:

```
document.head.appendChild(link);
HTMLImports.whenReady(callback):
```